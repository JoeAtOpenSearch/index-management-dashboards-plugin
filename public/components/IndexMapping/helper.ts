/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DiffableMappingsPropertiesObject, MappingsProperties, MappingsPropertiesObject } from "../../../models/interfaces";

export const transformObjectToArray = (obj: MappingsPropertiesObject): MappingsProperties => {
  return Object.entries(obj).map(([fieldName, fieldSettings]) => {
    const { properties, ...others } = fieldSettings;
    const payload: MappingsProperties[number] = {
      ...others,
      fieldName,
    };
    if (properties) {
      payload.properties = transformObjectToArray(properties);
    }
    return payload;
  });
};

export const transformArrayToObject = (array: MappingsProperties): MappingsPropertiesObject => {
  return array.reduce((total, current) => {
    const { fieldName, properties, ...others } = current;
    const payload: MappingsPropertiesObject[string] = {
      ...others,
    };
    if (properties) {
      payload.properties = transformArrayToObject(properties);
    }
    return {
      ...total,
      [current.fieldName]: payload,
    };
  }, {} as MappingsPropertiesObject);
};

export const transformArrayToDiffableObject = (array: MappingsProperties): DiffableMappingsPropertiesObject => {
  return array.reduce((total, current, index) => {
    const { properties, ...others } = current;
    const payload: MappingsPropertiesObject[string] = {
      ...others,
    };
    if (properties) {
      payload.properties = transformArrayToObject(properties);
    }
    return {
      ...total,
      [index]: payload,
    };
  }, {} as DiffableMappingsPropertiesObject);
};

export const countNodesInTree = (array: MappingsProperties) => {
  return array.reduce((total, current) => {
    total = total + 1;
    const { properties } = current;
    if (properties) {
      total = total + countNodesInTree(properties);
    }
    return total;
  }, 0);
};
