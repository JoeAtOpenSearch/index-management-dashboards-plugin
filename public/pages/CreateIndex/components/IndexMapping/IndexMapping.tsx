/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { forwardRef, useCallback, useState, Ref, useEffect, useRef } from "react";
import {
  EuiTreeView,
  EuiIcon,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiFieldText,
  EuiSelect,
  EuiTreeViewProps,
  EuiButton,
  EuiSpacer,
  EuiButtonGroup,
} from "@elastic/eui";
import { set, get } from "lodash";
import JSONEditor from "../../../../components/JSONEditor";
import { MappingsProperties, MappingsPropertiesObject } from "../../../../../models/interfaces";
import { INDEX_MAPPING_TYPES, INDEX_MAPPING_TYPES_WITH_CHILDREN } from "../../../../utils/constants";
import "./IndexMapping.scss";

export interface IndexMappingProps {
  value?: MappingsProperties;
  oldValue?: MappingsProperties;
  onChange: (value: IndexMappingProps["value"]) => void;
  isEdit?: boolean;
}

export interface IIndexDetailRef {
  validate: () => Promise<Boolean>;
}

export enum EDITOR_MODE {
  JSON = "JSON",
  VISUAL = "VISUAL",
}

interface IMappingLabel {
  value: MappingsProperties[number];
  onChange: (val: IMappingLabel["value"]) => void;
  // onFieldNameChange: (newFieldName: string, oldFieldName: string) => void;
  onAddSubField: () => void;
  disabled?: boolean;
}

const NEW_FIELD_PREFIX = "NAME_YOUR_FIELD";

const MappingLabel = ({ value, onChange, disabled, onAddSubField }: IMappingLabel) => {
  const { fieldName, ...fieldSettings } = value;
  const [fieldNameError, setFieldNameError] = useState("");
  const ref = useRef<any>(null);
  const type = fieldSettings.type ? fieldSettings.type : "object";
  const onFieldChange = useCallback(
    (k, v) => {
      const newValue = { ...value };
      set(newValue, k, v);
      onChange(newValue);
    },
    [value, onChange]
  );
  const [fieldNameState, setFieldNameState] = useState(fieldName);
  useEffect(() => {
    setFieldNameState(fieldNameState);
  }, [fieldName]);
  useEffect(() => {
    if (fieldNameState) {
      setFieldNameError("");
    }
  }, [fieldNameState]);
  return (
    <EuiFlexGroup onClick={(e) => e.stopPropagation()}>
      <EuiFlexItem style={{ width: 240 }} grow={false}>
        <EuiFormRow isInvalid={!!fieldNameError} error={fieldNameError} label="Field name" display="rowCompressed">
          <>
            <EuiFieldText
              inputRef={ref}
              disabled={disabled}
              compressed
              value={fieldNameState}
              onChange={(e) => setFieldNameState(e.target.value)}
              onFocus={() => {
                if (fieldNameState && fieldNameState.startsWith(NEW_FIELD_PREFIX)) {
                  setFieldNameState("");
                }
              }}
              onBlur={(e) => {
                if (!fieldNameState) {
                  setFieldNameError("Field name is required, please input");
                  e.target.focus();
                } else {
                  onFieldChange("fieldName", fieldNameState);
                }
              }}
            />
          </>
        </EuiFormRow>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiFormRow label="Field type" display="rowCompressed">
          <EuiSelect
            disabled={disabled}
            compressed
            value={type}
            onChange={(e) => onFieldChange("type", e.target.value)}
            options={INDEX_MAPPING_TYPES.map((item) => ({ text: item.label, value: item.label }))}
          />
        </EuiFormRow>
      </EuiFlexItem>
      {INDEX_MAPPING_TYPES_WITH_CHILDREN.includes(type) ? (
        <EuiFlexItem grow={false}>
          <EuiFormRow label="Extra parameters" display="rowCompressed">
            <span
              className="euiLink euiLink--primary"
              style={{ display: "flex", height: 32, alignItems: "center" }}
              onClick={(e) => {
                e.stopPropagation();
                onAddSubField();
              }}
            >
              Add a sub field
            </span>
          </EuiFormRow>
        </EuiFlexItem>
      ) : null}
    </EuiFlexGroup>
  );
};

const IndexMapping = ({ value, onChange, isEdit, oldValue }: IndexMappingProps, ref: Ref<IIndexDetailRef>) => {
  const treeRef = useRef<EuiTreeView>(null);
  const [editorMode, setEditorMode] = useState<EDITOR_MODE>(EDITOR_MODE.VISUAL);
  const addField = useCallback(
    (pos, fieldName) => {
      const newValue = [...(value || [])];
      const nowProperties = ((pos ? get(newValue, pos) : (newValue as MappingsProperties)) || []) as MappingsProperties;
      nowProperties.push({
        fieldName,
        type: "text",
      });
      onChange(newValue);
    },
    [onChange, value]
  );
  const transformValueToTreeItems = (formValue: IndexMappingProps["value"], pos: string = ""): EuiTreeViewProps["items"] => {
    return (formValue || []).map((item, index) => {
      const { fieldName, ...fieldSettings } = item;
      const id = [pos, index].filter((item) => item !== "").join(".properties.");
      const payload: EuiTreeViewProps["items"][number] = {
        label: (
          <MappingLabel
            disabled={isEdit && !!get(oldValue, id)}
            value={item}
            onChange={(val) => {
              const newValue = [...(value || [])];
              set(newValue, id, val);
              onChange(newValue);
            }}
            onAddSubField={() => {
              addField(`${id}.properties`, `${NEW_FIELD_PREFIX}-${Date.now()}`);
            }}
          />
        ),
        id,
        icon: <EuiIcon type="arrowRight" style={{ visibility: "hidden" }} />,
        iconWhenExpanded: <EuiIcon type="arrowDown" style={{ visibility: "hidden" }} />,
      };
      if (fieldSettings.properties) {
        (payload.icon = <EuiIcon type="arrowRight" />),
          (payload.iconWhenExpanded = <EuiIcon type="arrowDown" />),
          (payload.children = transformValueToTreeItems(fieldSettings.properties, id));
      }

      return payload;
    });
  };
  const transformedTreeItems = transformValueToTreeItems(value);
  return (
    <>
      <EuiSpacer />
      <EuiButtonGroup
        type="single"
        idSelected={editorMode as string}
        onChange={(id) => setEditorMode(id as EDITOR_MODE)}
        legend="Editor Type"
        options={[
          {
            label: "Visual Editor",
            id: EDITOR_MODE.VISUAL,
          },
          {
            label: "JSON Editor",
            id: EDITOR_MODE.JSON,
          },
        ]}
      />
      <EuiSpacer />
      {editorMode === EDITOR_MODE.VISUAL ? (
        <>
          {transformedTreeItems.length ? (
            <EuiTreeView
              key={JSON.stringify(value)}
              ref={treeRef}
              expandByDefault
              className="index-mapping-tree"
              aria-labelledby="label"
              items={transformValueToTreeItems(value)}
            />
          ) : (
            <p>You have no field mappings.</p>
          )}
          <EuiSpacer />
          <EuiButton onClick={() => addField("", `${NEW_FIELD_PREFIX}-${Date.now()}`)}>Add a field</EuiButton>
        </>
      ) : (
        <JSONEditor value={JSON.stringify(value || {}, null, 2)} onChange={(val) => onChange(JSON.parse(val))} />
      )}
    </>
  );
};

// @ts-ignore
export default forwardRef(IndexMapping);

export const transformObjectToArray = (obj: MappingsPropertiesObject): MappingsProperties => {
  return Object.entries(obj).map(([fieldName, fieldSettings]) => {
    const { properties, ...others } = fieldSettings;
    const payload: MappingsProperties[number] = {
      ...others,
      fieldName,
    };
    if (properties) {
      payload.properties = transformObjectToArray(properties);
    }
    return payload;
  });
};

export const transformArrayToObject = (array: MappingsProperties): MappingsPropertiesObject => {
  return array.reduce((total, current) => {
    const { fieldName, properties, ...others } = current;
    const payload: MappingsPropertiesObject[string] = {
      ...others,
    };
    if (properties) {
      payload.properties = transformArrayToObject(properties);
    }
    return {
      ...total,
      [current.fieldName]: payload,
    };
  }, {} as MappingsPropertiesObject);
};
