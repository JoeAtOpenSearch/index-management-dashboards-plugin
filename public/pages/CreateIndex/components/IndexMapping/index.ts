/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexMapping from "./IndexMapping";

export default IndexMapping;
