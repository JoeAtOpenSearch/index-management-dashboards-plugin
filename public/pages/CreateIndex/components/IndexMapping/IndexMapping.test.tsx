/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import IndexMapping from "./IndexMapping";

describe("<IndexMapping /> spec", () => {
  it("renders the component", () => {
    const { container } = render(<IndexMapping onChange={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("render mappings with object type", () => {
    const { container } = render(
      <IndexMapping
        onChange={() => {}}
        value={[{ fieldName: "object", type: "object", properties: [{ fieldName: "text", type: "text" }] }]}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
