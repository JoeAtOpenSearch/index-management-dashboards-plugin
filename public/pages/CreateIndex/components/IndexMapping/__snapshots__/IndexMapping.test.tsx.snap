// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexMapping /> spec render mappings with object type 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <fieldset
    class="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
  >
    <legend
      class="euiScreenReaderOnly"
    >
      Editor Type
    </legend>
    <div
      class="euiButtonGroup__buttons"
    >
      <label
        class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected"
        for="some_html_id"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text euiButtonGroupButton__textShift"
            data-text="Visual Editor"
            title="Visual Editor"
          >
            <input
              checked=""
              class="euiScreenReaderOnly"
              data-test-subj="VISUAL"
              id="some_html_id"
              name="some_html_id"
              type="radio"
              value=""
            />
            Visual Editor
          </span>
        </span>
      </label>
      <label
        class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small"
        for="some_html_id"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text euiButtonGroupButton__textShift"
            data-text="JSON Editor"
            title="JSON Editor"
          >
            <input
              class="euiScreenReaderOnly"
              data-test-subj="JSON"
              id="some_html_id"
              name="some_html_id"
              type="radio"
              value=""
            />
            JSON Editor
          </span>
        </span>
      </label>
    </div>
  </fieldset>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiText euiText--medium euiTreeView__wrapper"
  >
    <p
      class="euiScreenReaderOnly"
      id="some_html_id--instruction"
    >
      You can quickly navigate this list using arrow keys.
    </p>
    <ul
      aria-describedby="some_html_id--instruction"
      aria-labelledby="label"
      class="euiTreeView index-mapping-tree"
      id="some_html_id"
    >
      <li
        class="euiTreeView__node euiTreeView__node--expanded"
      >
        <button
          aria-controls="some_html_id"
          aria-expanded="true"
          class="euiTreeView__nodeInner"
          data-test-subj="euiTreeViewButton-some_html_id"
          id="0"
        >
          <span
            class="euiTreeView__iconWrapper"
          >
            EuiIconMock
          </span>
          <span
            class="euiTreeView__nodeLabel"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
                style="width: 240px;"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      Field name
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout euiFormControlLayout--compressed"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          class="euiFieldText euiFieldText--compressed"
                          type="text"
                          value="object"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      Field type
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout euiFormControlLayout--compressed"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <select
                          class="euiSelect euiSelect--compressed"
                          id="some_html_id"
                        >
                          <option
                            value="null"
                          >
                            null
                          </option>
                          <option
                            value="boolean"
                          >
                            boolean
                          </option>
                          <option
                            value="float"
                          >
                            float
                          </option>
                          <option
                            value="double"
                          >
                            double
                          </option>
                          <option
                            value="integer"
                          >
                            integer
                          </option>
                          <option
                            value="text"
                          >
                            text
                          </option>
                          <option
                            value="keyword"
                          >
                            keyword
                          </option>
                          <option
                            value="date"
                          >
                            date
                          </option>
                          <option
                            value="object"
                          >
                            object
                          </option>
                        </select>
                        <div
                          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                        >
                          <span
                            class="euiFormControlLayoutCustomIcon"
                          >
                            EuiIconMock
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      actions
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      id="some_html_id"
                      style="display: flex; height: 32px; align-items: center;"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-label="Add a sub field"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--medium"
                          type="button"
                        >
                          EuiIconMock
                        </button>
                      </span>
                      <span
                        class="euiToolTipAnchor"
                      >
                        <button
                          aria-label="Delete current field"
                          class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--medium"
                          type="button"
                        >
                          EuiIconMock
                        </button>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
        </button>
        <div
          id="some_html_id"
        >
          <div
            class="euiText euiText--medium euiTreeView__wrapper"
          >
            <ul
              aria-labelledby="0 label"
              class="euiTreeView"
            >
              <li
                class="euiTreeView__node"
              >
                <button
                  aria-controls="some_html_id"
                  aria-expanded="false"
                  class="euiTreeView__nodeInner"
                  data-test-subj="euiTreeViewButton-some_html_id"
                  id="0.properties.0"
                >
                  <span
                    class="euiTreeView__iconWrapper"
                  >
                    EuiIconMock
                  </span>
                  <span
                    class="euiTreeView__nodeLabel"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                        style="width: 240px;"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              aria-invalid="false"
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              Field name
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <div
                              class="euiFormControlLayout euiFormControlLayout--compressed"
                            >
                              <div
                                class="euiFormControlLayout__childrenWrapper"
                              >
                                <input
                                  class="euiFieldText euiFieldText--compressed"
                                  type="text"
                                  value="text"
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              Field type
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <div
                              class="euiFormControlLayout euiFormControlLayout--compressed"
                            >
                              <div
                                class="euiFormControlLayout__childrenWrapper"
                              >
                                <select
                                  class="euiSelect euiSelect--compressed"
                                  id="some_html_id"
                                >
                                  <option
                                    value="null"
                                  >
                                    null
                                  </option>
                                  <option
                                    value="boolean"
                                  >
                                    boolean
                                  </option>
                                  <option
                                    value="float"
                                  >
                                    float
                                  </option>
                                  <option
                                    value="double"
                                  >
                                    double
                                  </option>
                                  <option
                                    value="integer"
                                  >
                                    integer
                                  </option>
                                  <option
                                    value="text"
                                  >
                                    text
                                  </option>
                                  <option
                                    value="keyword"
                                  >
                                    keyword
                                  </option>
                                  <option
                                    value="date"
                                  >
                                    date
                                  </option>
                                  <option
                                    value="object"
                                  >
                                    object
                                  </option>
                                </select>
                                <div
                                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                                >
                                  <span
                                    class="euiFormControlLayoutCustomIcon"
                                  >
                                    EuiIconMock
                                  </span>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              actions
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <div
                              id="some_html_id"
                              style="display: flex; height: 32px; align-items: center;"
                            >
                              <span
                                class="euiToolTipAnchor"
                              >
                                <button
                                  aria-label="Delete current field"
                                  class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--medium"
                                  type="button"
                                >
                                  EuiIconMock
                                </button>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </span>
                </button>
                <div
                  id="some_html_id"
                />
              </li>
            </ul>
          </div>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <button
    class="euiButton euiButton--primary"
    type="button"
  >
    <span
      class="euiButtonContent euiButton__content"
    >
      <span
        class="euiButton__text"
      >
        Add a field
      </span>
    </span>
  </button>
</div>
`;

exports[`<IndexMapping /> spec renders the component 1`] = `
<div
  class="euiSpacer euiSpacer--l"
/>
`;
